### PR TITLE
feat(web): VisualKeyboard layout-reflow optimization 🪠

### PR DESCRIPTION
**Pull request by jahorton**
Now for some of the "meat" of the OSK optimization - this PR heavily reduces `VisualKeyboard` layout reflow, batching similar operations (read vs write) as much as possible at all times.  Outside of a few knock-on effects for key and gesture previews that needed resolving, all changes are restricted to the scope of `VisualKeyboard.refreshLayout()`, the methods it calls, and the values it requests.

This PR has been spun off of an older state of https://github.com/keymanapp/keyman/pull/11140.

@keymanapp-test-bot skip

All testing reasonable to do here is also reasonable to do on #11140.

### CODE REVIEW

**Review by ermshiperete — APPROVED** [review]

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.307-beta](https://keyman.com/downloads/releases/beta/17.0.307)

File: web/src/engine/osk/src/keyboard-layout/oskBaseKey.ts [hunk 1]
```diff
@@ -197,29 +196,23 @@ export default class OSKBaseKey extends OSKKey {
     this.btn.replaceChild(this.preview, oldPreview);
   }
 
-  public refreshLayout(vkbd: VisualKeyboard) {
-    let key = this.spec as ActiveKey;
-    this.square.style.width = vkbd.layoutWidth.scaledBy(key.proportionalWidth).styleString;
-    this.square.style.marginLeft = vkbd.layoutWidth.scaledBy(key.proportionalPad).styleString;
-    this.btn.style.width = vkbd.usesFixedWidthScaling ? this.square.style.width : '100%';
+  public refreshLayout(layoutParams: KeyLayoutParams) {
+    const keyTextClosure = super.refreshLayout(layoutParams);  // key labels in particular.
```
Comment (mcdurdin) [inline web/src/engine/osk/src/keyboard-layout/oskBaseKey.ts:200]: Why 'keyTextClosure'? It's really unclear what that does! Perhaps we should call it `superFinalization` to clarify that it's more work that super.refreshlayout has to do once this function is finished?

File: web/src/engine/osk/src/keyboard-layout/oskKey.ts [hunk 2]
```diff
@@ -310,47 +294,73 @@ export default abstract class OSKKey {
...
     // space bar may not define the text span!
     if(this.label) {
       if(!this.label.classList.contains('kmw-spacebar-caption')) {
         // Do not use `this.keyText` - it holds *___* codes for special keys, not the actual glyph!
         const keyCapText = this.label.textContent;
-        this.label.style.fontSize = this.getIdealFontSize(vkbd, keyCapText, this.btn.style);
+        const fontSize = this.getIdealFontSize(keyCapText, layoutParams);
+        return () => {
+          this.label.style.fontSize = fontSize.styleString;
+        };
```
Comment (mcdurdin) [inline web/src/engine/osk/src/keyboard-layout/oskKey.ts:353]: This is kinda ugly and it's unclear why this is postponed and how it gets called

File: web/src/engine/osk/src/keyboard-layout/oskKey.ts [hunk 3]
```diff
@@ -310,47 +294,73 @@ export default abstract class OSKKey {
...
+  public resetFontPrecalc() {
+    this._fontFamily = undefined;
+    this._fontSize = undefined;
+    this.label.style.fontSize = '';
+  }
+
+  public refreshLayout(layoutParams: KeyLayoutParams) {
+    // Avoid doing any font-size related calculations if there's no text to display.
+    if(this.spec.sp == ButtonClasses.spacer || this.spec.sp == ButtonClasses.blank) {
+      return () => {};
```
Comment (mcdurdin) [inline web/src/engine/osk/src/keyboard-layout/oskKey.ts:317]: It'd be better to `return null` and then in the caller, use `closure?.()` to make the call: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining#optional_chaining_with_function_calls

File: web/src/engine/osk/src/keyboard-layout/oskLayer.ts [hunk 4]
```diff
@@ -132,42 +136,63 @@ export default class OSKLayer {
...
+    for(const row of this.rows) {
+      const batchedUpdates = row.refreshKeyLayouts(layoutParams);
+      rowKeyClosures.push(batchedUpdates);
     }
+
+    // After row layout properties have been updated, _then_ update key internals.
+    // Doing this separately like this helps to reduce layout reflow.
+    spacebarTextClosure();
+    rowClosures.forEach((closure) => closure());
+    rowKeyClosures.forEach((closure) => closure());
```
Comment (mcdurdin) [inline web/src/engine/osk/src/keyboard-layout/oskLayer.ts:196]: I'm really not keen on all this passing around of closures. It's really hard to follow what is happening when. It shouldn't be necessary with a good structure to the code.

File: web/src/engine/osk/src/keyboard-layout/oskLayer.ts [hunk 5]
```diff
@@ -132,42 +136,63 @@ export default class OSKLayer {
...
       }
     }
     catch (ex) { }
   }
 
-  public refreshLayout(vkbd: VisualKeyboard, layoutParams: LayerLayoutParams) {
-    const layerHeight = layoutParams.keyboardHeight;
-    this.showLanguage(layoutParams.spacebarText);
-
+  public refreshLayout(layoutParams: LayerLayoutParams) {
```
Comment (mcdurdin) [inline web/src/engine/osk/src/keyboard-layout/oskLayer.ts:151]: So this implementation of `refreshLayout` doesn't return a closure but other ones do. It would be good if this was clear without having to search for `return` statements.

**Review by mcdurdin — COMMENTED** [review]
I am not comfortable with this change. We may have to live with it for 17.0 because we are running out of time, but the code flow is extremely obtuse due to the extensive use of closures, and I think it will be very hard to maintain.

I think we should be able to simplify some of these patterns dramatically with a bit of careful design, what do you think?

File: web/src/engine/osk/src/keyboard-layout/oskBaseKey.ts (hunk 1 above)
Comment (jahorton) [reply web/src/engine/osk/src/keyboard-layout/oskBaseKey.ts:200]: > Why 'keyTextClosure'?

Because it's a closure to finalize/apply layout changes directly tied to key text as seen in OSKKey.  

> Perhaps we should call it `superFinalization` [...]

If I swapped to `superFinalization` or something else unrelated to the code-contents of the closure... well, I suppose I can always just add a comment about what the closure's actual effects are.

File: web/src/engine/osk/src/keyboard-layout/oskKey.ts (hunk 2 above)
Comment (jahorton) [reply web/src/engine/osk/src/keyboard-layout/oskKey.ts:353]: This is the other side of that `keyTextClosure` you were asking about - its actual definition.

Admittedly, yeah, the `keyTextClosure` aspect is probably the least-clear thing in this PR.  It got tricky to forward well due to the inheritance structures involved.

File: web/src/engine/osk/src/keyboard-layout/oskLayer.ts (hunk 4 above)
Comment (jahorton) [reply web/src/engine/osk/src/keyboard-layout/oskLayer.ts:196]: Huh.

To me, I felt that this made it extremely clear that we've deferred and essentially batched all operations that could affect DOM layout (and thus, could trigger layout reflow) to this one place.  (Granted, maybe I should have added documentation of this detail to the lower-level closure-building calls...)  All of the closures rely on precomputed values, some of which do refer to existing layout (when prerequisite values are uncached).  By precomputing all values first, without any edits... and then triggering those edits later... that's how we prevent layout reflow here.

Now that all precomputations are complete, we're safe to apply all of them sequentially without any reference to layout or style information.  Use of closures automatically caches the precomputations with manually-defined code, which is a major reason I used the pattern.  (Rather than manually tracking the values, we can just leverage existing language features to do that heavy lifting.)

---

Granted, maybe some extra comments could make this clearer - whether here or on the lower-level `refresh` methods in regard to why the returned closures exist.  What would make this clearer for you?  Alternatively... what alternate "good structure" would you suggest?

**Review by jahorton — COMMENTED** [review]
> I think we should be able to simplify some of these patterns dramatically with a bit of careful design, what do you think?

I can only really see it working if we promote those closures to class methods... and manually construct the value-tracking objects that we get for free by using closures.  I did look at that pattern initially, but it started getting a bit too gnarly for my tastes for the cases where key-style values weren't yet cached.

----

Then again, maybe having a "try-cache" method (sorry re: the horrible pun) before the precomputation pass could work?  (As in, "fetch key font-style info if not yet cached and cacheable".)

Waaaaaaait.  That (the "try-cache") might actually work better.  We'd need to test, but I _think_ it'd be possible to run that _specific_ call earlier in the `VisualKeyboard.refreshLayout` method - before it runs its `getComputedStyle` calls.  If so, I believe we'd be able to eliminate a major layout-reflow source I previously thought we couldn't.

(Reasoning:  the font-size computation either picks up a fixed font-size - in which case, changing the base font-scaling doesn't matter - or it picks up a relative one... and still successfully computes the _local, key-specific_ scaling factor just fine.)

**Comment by mcdurdin** [comment]
> Granted, maybe some extra comments could make this clearer - whether here or on the lower-level refresh methods in regard to why the returned closures exist. What would make this clearer for you? Alternatively... what alternate "good structure" would you suggest?

Some extra comments would definitely help. The problem with the term `closure` is that I have to go and find the definition of it to figure out what it is actually doing.

https://github.com/keymanapp/keyman/blob/573ffd8924b5aa6ada60a651cd34bd5d47f4eecd/web/src/engine/osk/src/keyboard-layout/oskLayer.ts#L195-L196
